flatten operation (resnet50)

migeed-z opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #61790
* #61776
* #61747
* #61310
* **#61265**

Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

migeed-z committed:
flatten operation (resnet50)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"




[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626383).

jamesr66a commented on test/fx/test_gradual_type.py:
```diff
@@ -551,7 +551,7 @@ def forward(self, x: TensorType((2, 2, 4, 5))):
                 assert torch.Size(n.type.__args__) == B.forward(torch.rand(2, 2, 4, 5)).size()
 
-    def test_type_check_conv2D_and_maxpool2d(self):
+    def test_type_check_conv2D__maxpool2d_flatten(self):
```
nit: double `_`

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -374,6 +372,51 @@ def adaptiveavgpool2d_inference_rule(n: Node, module_instance):
+        assert isinstance(n.args[2], int)
+        end_dim = n.args[2]
+
+    if n.args[0] == Dyn:
```
Should this be `n.args[0].type == Dyn`?

Also can you test this case

jamesr66a commented on torch/fx/experimental/graph_gradual_typechecker.py:
```diff
@@ -374,6 +372,51 @@ def adaptiveavgpool2d_inference_rule(n: Node, module_instance):
+def flatten_check(typ, start_dim, end_dim):
+    l = len(typ.__args__)
+
+    if 0 <= start_dim <= (l - 1) and end_dim == -1 or 0 <= end_dim <= (l - 1) and start_dim < end_dim:
```
Can you describe what this is checking?

I suggest pre-normalizing the dimensions at the very beginning so we don't have to have extra checks for `-1`, e.g.:

```
def flatten_check(typ, start_dim, end_dim):
    l = len(typ.__args__)

    start_dim = (l - 1) if start_dim == -1 else start_dim
    end_dim = (l - 1) if end_dim == -1 else end_dim
```

Then I believe this check should just be
```
if start_dim <= end_dim:
```

migeed-z replied:
that makes sense. I was also checking if the user gave a non-existent index, for example a tensor of size 3 and they give index 10. If so I think we should raise a type error too.

migeed-z replied:
also  I used l instead of (l - 1)  because of the list comprehensions. For example [1,2,3,5,6][:5] =  [1,2,3,5,6]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626383).

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z committed:
Update on "flatten operation (resnet50)"


Differential Revision: [D29626383](https://our.internmc.facebook.com/intern/diff/D29626383)

[ghstack-poisoned]

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626383).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626383).

migeed-z commented:
@migeed-z has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29626383).
